Make WebCore::Site distinct from HashTableEmptyValueType

achristensen07 opened the pull request:
#### a6c5148743e3513a8a161e3955811b2046771327
<pre>
Make WebCore::Site distinct from HashTableEmptyValueType
<a href="https://bugs.webkit.org/show_bug.cgi?id=293155">https://bugs.webkit.org/show_bug.cgi?id=293155</a>

Reviewed by Sihui Liu.

Add an invalidProtocol for Sites constructed with empty protocols
similar to RegistrableDomain&apos;s use of nullOrigin, and for the same reason.

* Source/WebCore/platform/Site.cpp:
(WebCore::invalidProtocol):
(WebCore::nonEmptyProtocol):
(WebCore::Site::Site):
(WebCore::Site::protocol const):
* Source/WebCore/platform/Site.h:
(WebCore::Site::protocol const): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/RegistrableDomain.cpp:
(TestWebKitAPI::TEST(Site, EmptyMapKey)):

Canonical link: <a href="https://commits.webkit.org/295063@main">https://commits.webkit.org/295063@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ba654e4be129667c1d78b205c294c2d48cf7247

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14013 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109185 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54654 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106029 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24051 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32237 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78996 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93807 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59324 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18468 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11858 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54017 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111569 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31145 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22950 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88007 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31509 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90007 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87664 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32552 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10294 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25539 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16880 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31074 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36385 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30867 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34204 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Tools/TestWebKitAPI/Tests/WebCore/RegistrableDomain.cpp:
```diff
@@ -99,4 +100,20 @@ TEST(RegistrableDomain, UncheckedCreateFromHost)
+    map.set(Site(URL("!@#$%^&*NotARealURL)(*&^"_s)), 2);
+    map.set(Site(SecurityOriginData(String(), String(), std::nullopt)), 3);
+    map.set(Site(String(), RegistrableDomain::fromRawString(String())), 4);
+    EXPECT_EQ(map.size(), 1u);
```
Can you add map size check after each set (to check if it's overwrite or failing to set)?

szewai commented on Tools/TestWebKitAPI/Tests/WebCore/RegistrableDomain.cpp:
```diff
@@ -99,4 +100,20 @@ TEST(RegistrableDomain, UncheckedCreateFromHost)
+        EXPECT_TRUE(site.isEmpty());
+        EXPECT_EQ(value, 4);
+    }
+    map.set(Site(String(), RegistrableDomain()), 5);
```
I feel it's a bit counterintuitive that `Site(String(), RegistrableDomain::fromRawString(String()))` and `Site(String(), RegistrableDomain())` are not mapped to the same entry when `Site::isEmpty()` returns true for both.
If this is expected, I think we need to add a `EXPECT_TRUE(site.isEmpty());` here

achristensen07 committed:
Make WebCore::Site distinct from HashTableEmptyValueType
https://bugs.webkit.org/show_bug.cgi?id=293155

Reviewed by Sihui Liu.

Add an invalidProtocol for Sites constructed with empty protocols
similar to RegistrableDomain's use of nullOrigin, and for the same reason.

* Source/WebCore/platform/Site.cpp:
(WebCore::invalidProtocol):
(WebCore::nonEmptyProtocol):
(WebCore::Site::Site):
(WebCore::Site::protocol const):
* Source/WebCore/platform/Site.h:
(WebCore::Site::protocol const): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/RegistrableDomain.cpp:
(TestWebKitAPI::TEST(Site, EmptyMapKey)):

Canonical link: https://commits.webkit.org/295063@main
